data-prepper-plugins/kinesis-source/src/main/java/org/opensearch/dataprepper/plugins/kinesis/source/processor/KinesisRecordProcessor.java

```diff
@@ -0,0 +1,270 @@
+                LOG.debug("acknowledgements received for stream: {}, shardId: {}", kinesisStreamName, kinesisShardId);
+            } else {
+                acknowledgementSetFailures.increment();
+                LOG.debug("acknowledgements received with false for stream: {}, shardId: {}", kinesisStreamName, kinesisShardId);
+            }
+        }, ACKNOWLEDGEMENT_SET_TIMEOUT);
+    }
+
+    @Override
+    public void processRecords(ProcessRecordsInput processRecordsInput) {
+        try {
+            Optional<AcknowledgementSet> acknowledgementSetOpt = Optional.empty();
+            boolean acknowledgementsEnabled = kinesisSourceConfig.isAcknowledgments();
+            ExtendedSequenceNumber extendedSequenceNumber = getLatestSequenceNumberFromInput(processRecordsInput);
+            if (acknowledgementsEnabled) {
+                acknowledgementSetOpt = Optional.of(createAcknowledgmentSet(processRecordsInput, extendedSequenceNumber));
+            }
+
+            // Track the records for checkpoint purpose
+            kinesisCheckpointerTracker.addRecordForCheckpoint(extendedSequenceNumber, processRecordsInput.checkpointer());
+            List<Record<Event>> records = kinesisRecordConverter.convert(processRecordsInput.records());
+
+            int eventCount = 0;
+            for (Record<Event> record: records) {
+                Event event = record.getData();
+                acknowledgementSetOpt.ifPresent(acknowledgementSet -> acknowledgementSet.add(event));
+                EventMetadata eventMetadata = event.getMetadata();
+                eventMetadata.setAttribute(MetadataKeyAttributes.KINESIS_STREAM_NAME_METADATA_ATTRIBUTE,
+                        streamIdentifier.streamName().toLowerCase());
+                bufferAccumulator.add(record);
+                eventCount++;
+            }
+
+            // Flush buffer at the end
+            bufferAccumulator.flush();
+            recordsProcessed.increment(eventCount);
+
+            // If acks are not enabled, mark the sequence number for checkpoint
+            if (!acknowledgementsEnabled) {
+                kinesisCheckpointerTracker.markSequenceNumberForCheckpoint(extendedSequenceNumber);
+            }
+
+            LOG.debug("Number of Records {} written for stream: {}, shardId: {} to buffer: {}", eventCount, streamIdentifier.streamName(), kinesisShardId, records.size());
+
+            acknowledgementSetOpt.ifPresent(AcknowledgementSet::complete);
+
+            // Checkpoint for shard
+            if (!acknowledgementsEnabled && (System.currentTimeMillis() - lastCheckpointTimeInMillis >= checkpointInterval.toMillis())) {
+                doCheckpoint();
+            }
+        } catch (Exception ex) {
+            recordProcessingErrors.increment();
+            LOG.error("Failed writing shard data to buffer: ", ex);
+        }
+    }
+
+    @Override
+    public void leaseLost(LeaseLostInput leaseLostInput) {
+        LOG.debug("Lease Lost");
```

Comment:
Is this worth having a metric for or does kinesis already track it? Why does this happen?

Reply:
KCL already tracks this through a metric. This can happen if the current worker holding the lease is evicted by another worker which acquires the lease for the specific shard.

Reply:
This is already tracked by KCL. This happens when the current worker holding the lease is evicted due to timeout by another worker which just acquired the lease for the specific shard.